doc/reference/std/instance.md

```diff
@@ -0,0 +1,269 @@
+  p ‘bind‘ f = \inp -> concat [f v out | (v,out) <- p inp]
+```
+
+A translation to gerbil uses **Interface Passing Style**. We'll take a struct as the interface object.
+
+```scheme
+(defstruct parser ())
+```
+
+If using the builtin `defmethod` it's not that far off the functional version.
+
+```scheme
+(import :std/srfi/1)
+;; return :: a -> Parser a
+(defmethod {return parser} (lambda (_ v) (lambda (inp) [[v inp ...]])))
+;; bind :: Parser a -> (a -> Parser b) -> Parser b
+(defmethod {bind parser}
+ (lambda (_ p f)
+   (lambda (inp)
+     (append-map (cut match <> ([v . out] ((f v) out)))
+                (p inp)))))
+```
+
+But we can do better. Introducting the `instance` syntax.
+
+```scheme
+(instance Monad parser
+ ;; return :: a -> Parser a
+ ((return v) (lambda (inp) [[v inp ...]]))
+ ;; bind :: Parser a -> (a -> Parser b) -> Parser b
+ ((bind p f)
+  (lambda (inp) (append-map (cut match <> ([v . out] ((f v) out))) (p inp)))))
+
+```
+
+
+## The `instance` Syntax
+
+```scheme
+(instance Interface Klass
+ method-def ...
+ [rebind: <boolean>])
+
+<Interface>:
+ <symbol> ;; The name of the interface
+ (instance Interface) ;; both <symbol>'s, the name and instance of interface.
+ <boolean> ;; The interface is unused/blank
+
+<Klass>:
+ <symbol> ;; The name of the class we are defining methods on
+ (self Klass) ;; both <symbol>'s, the name of the class and self thereof.
+
+<method-def>:
+ ((name . args) body ...) ;; The definition of a method. Note the
+                          ;; object is inferred so not in the lambda list.
+
+<rebind>:
+ <boolean> ;; Default: #f ; Error if not true and any method already
+         ;; been bound.. Otherwise rebind the klass.
+```
+
+The idea here is simply to reduce typing while putting things in a concise container.
+
+
+## With an `interface` but no object
+
+For example let's start with a bigger `Monad`.
+
+We've seen a Monad Interface.
+
+By default Haskell has one more function that simply "combinds" two monadic functions.
+
+```haskell
+class  Monad m  where
+  bind  :: m a -> (a -> m b) -> m b
+  seq   :: m a -> m b -> m b
+  return :: a -> m a
+
+      -- Minimal complete definition:
+      --      (>>=), return
+  m >> k  =  m >>= \_ -> k
+```
+
+We can add that to our interface:
+
+```scheme
+(interface Monad
+  (return a) #;| -> m a |
+  (bind m a) #;| -> (a -> m b) -> m b |
+  (seq ma mb) #;| m a -> m b -> m b |)
+```
+
+And make the `:t` class into an identity monad with a minimal complete definition.
+
+```scheme
+(instance (m Monad) :t
+ ((return a) a)
+ ((bind ma f) (f ma))
+ ((seq ma mb) (m.bind ma (lambda _ mb))))
+```
+
+The `((return a) a)` form expands to:
+
+```scheme
+(defmethod {return :t} (lambda (self a) (using (m self : Monad) a)) rebind: #f)
+```
+
+A binding of `self` is just made up and hygenic as it's not used.
+
+And the `seq` expansion becomes obvious as well.
+
+```scheme
+(defmethod {seq :t}
+  (lambda (self ma mb) (using (m self : Monad) (m.bind ma (lambda _ mb))))
```

Comment:
rely on the implicit annotation here, yhis will likely interfere with specialization.

Reply:
hrm no, it is fine i think.